refactor to lodash/core

mweitzel opened the pull request:
Hello,

I refactored the project to exclusively use `lodash/core` instead of `lodash`, including a cloned local instance of `lodash-deeper`.

This is nowhere near ready to merge, but I was able to to some numbers. I just wanted to put this here and open the floor for discussion.

Cheers,
Matthew

![screen shot 2016-07-15 at 2 54 04 pm](https://cloud.githubusercontent.com/assets/318925/16885460/a42be472-4a9c-11e6-9c30-c2718f0739c7.png)

searls commented:
This is VERY interesting Matthew, thanks for taking the initiative. I'll dive in soon.

searls commented:
So @mweitzel this is passing in CI, what tells you this isn't ready to merge? Still some WIP references to clear out?

mweitzel commented:
There are a few `_.isNil` calls I've swapped out for `!=`, which would have been fine if required as `_isNil`, as it has no dependencies. But I didn't notice until about half way through, so I kept gutting them to stay consistent. It could go either way, and should really come down to the maintainer's discretion.

Things like `omit`, however, pull in all sorts of other files and would have significant impact.

The biggest thing I noticed was the lack of support for nested/deep traversal. Where you have to provide callbacks or simple strings as predicates instead of `[key, value]`, or `chain.into.me` predicates to match.

mweitzel commented:
@searls a few things:

[deeper](https://raw.githubusercontent.com/mweitzel/teenytest/6138cec33cc57620272bc6d2dd0e2875797ea971/lodash-deeper-core.js) is an issue. We either should cut the dep from `package.json` or patch and bump the version number.

The omit modules has no testing. I'd want to either add a test or few to that or use `_omit` after ensuring it truly doesn't raise overhead.

I wanted to get this up before I left for the weekend, so I did rush a touch and would definitely want another set of eyes.

searls commented:
Seems like a good start. I think the next thing I'd really like is a set of benchmarks against the CLI committed to the repo so we can be fact-based about performance PRs

mweitzel commented:
@searls sounds great. I assume you'd want that as a separate pr?

mweitzel commented:
crossreference to testdouble/lodash-deeper#1

That PR allows consuming libraries (like this one) to explicitly state they want only to depend on `lodash/core`.

If that merges and is npm published, I can bump the dependency in `package.json` and nix the cloned/edited file in this project's root directory.

searls commented:
Yeah for sure.

mweitzel commented:
OK, I gave it another pass and am pretty happy with where it stands. Much better than it looked on Friday.

mweitzel commented:
Thanks again for pushing the lodash-deeper patch through.

searls commented:
Paging @jasonkarns to review this if he gets a chance since he had a similar PR

I think we should wait to merge until we have data on benchmark perf

jasonkarns commented on lib/build-test-helper.js:
```diff
@@ -16,7 +16,7 @@ module.exports = function (helperPath, cwd) {
 function buildUserHelper (helperPath, cwd) {
   if (helperPath && fs.existsSync(helperPath)) {
-    return _.assign({}, require(path.resolve(cwd, helperPath)), {
+    return _.assignIn({}, require(path.resolve(cwd, helperPath)), {
```
Are we sure we want to assign inherited properties as well? (I don't know the answer. assuming assignIn is sufficient, but it _could_ be slower)

jasonkarns commented:
I admire the work, but I'm not yet convinced that the performance will be worth the code readability changes. Remembering strict null/undefined checks is one thing, but there are an awful lot of object traversal and other things that I'm not sure is worth sacrificing.

Unless the benchmarks prove a drastic speed improvement, my default position is to just keep big lodash.

However, if it proves fast enough, another option would be to copy/paste the useful (but omitted from /core) utility methods out of lodash into a single lib file.

mweitzel commented:
I very much like the idea of a helper module, to bridge the gap between lodash/core and good readability. Splitting functions and modules from library functions and modules. Though, I'm pretty sure object traversal is a huge part of big lodash's girth.

Using [this script](https://gist.github.com/mweitzel/bbbdc3cc3b9d09080764bce59c28d42c) to compare `require('./index')` across branches, lodash-core and master. It measures explicitly the library require time, from within node so there's muddied by node's system time. Here's what I see:

```
$ ./perf/require-time index.js 50
comparing branches lodash-core and master..

require times on branch lodash-core:
1   77ms  *
1   75ms  *
4   74ms  ****
1   73ms  *
2   72ms  **
1   71ms  *
7   70ms  *******
18  69ms  ******************
14  68ms  **************
1   67ms  *

Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.

require times on branch master:
1   142ms  *
1   135ms  *
1   132ms  *
1   126ms  *
2   125ms  **
1   122ms  *
3   121ms  ***
1   120ms  *
3   119ms  ***
2   118ms  **
5   117ms  *****
6   116ms  ******
11  115ms  ***********
11  114ms  ***********
1   113ms  *

Switched to branch 'lodash-core'
```

I'm not sure of the performance of my machine, and would be curious to see it run on others. If my machine is just slow and the difference is negligible on most people's hardware, I completely understand your position.

I'd also love some benchmarks against the CLI, but I'm not sure that's appropriate for this already extensive PR.

jasonkarns commented:
This seems to have languished since the refactor, but it might also be worth considering just using lodash's CLI to generate a custom build with the methods we're using. I'd be curious to see how that performs. (Essentially being an easy-to-generate helper with only the funcs we use.)

https://www.npmjs.com/package/lodash-cli

searls commented:
Sounds good

> On Sep 23, 2017, at 00:25, Jason Karns <notifications@github.com> wrote:
> 
> This seems to have languished since the refactor, but it might also be worth considering just using lodash's CLI to generate a custom build with the methods we're using. I'd be curious to see how that performs. (Essentially being an easy-to-generate helper with only the funcs we use.)
> 
> https://www.npmjs.com/package/lodash-cli
> 
> ―
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
